pkg/virt-controller/watch/snapshot/source.go

```diff
@@ -0,0 +1,236 @@
+		return false, err
+	}
+
+	exists, err := s.controller.checkVMIRunning(s.vm)
+	if err != nil {
+		return false, err
+	}
+
+	if !vmRunning && !exists {
+		pvcNames := s.pvcNames()
+		pods, err := podsUsingPVCs(s.controller.PodInformer, s.vm.Namespace, pvcNames)
+		if err != nil {
+			return false, err
+		}
+
+		if len(pods) > 0 {
+			log.Log.V(3).Infof("Vm is offline but %d pods using PVCs %+v", len(pods), pvcNames)
+			return false, nil
+		}
+	}
+
+	if s.vm.Status.SnapshotInProgress != nil && *s.vm.Status.SnapshotInProgress != s.snapshot.Name {
+		log.Log.V(3).Infof("Snapshot %s in progress", *s.vm.Status.SnapshotInProgress)
+		return false, nil
+	}
+
+	log.Log.Infof("Adding VM snapshot finalizer to %s", s.vm.Name)
+
+	vmCopy := s.vm.DeepCopy()
+
+	if vmCopy.Status.SnapshotInProgress == nil {
+		vmCopy.Status.SnapshotInProgress = &s.snapshot.Name
+		// unfortunately, status updater does not return the updated resource
+		// but the controller is watching VMs so will get notified
+		return false, s.controller.vmStatusUpdater.UpdateStatus(vmCopy)
+	}
+
+	if !controller.HasFinalizer(vmCopy, sourceFinalizer) {
+		controller.AddFinalizer(vmCopy, sourceFinalizer)
+		_, err = s.controller.Client.VirtualMachine(vmCopy.Namespace).Update(vmCopy)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func (s *vmSnapshotSource) Unlock() (bool, error) {
+	if s.vm.Status.SnapshotInProgress == nil || *s.vm.Status.SnapshotInProgress != s.snapshot.Name {
+		return false, nil
+	}
+
+	var err error
+	vmCopy := s.vm.DeepCopy()
+
+	if controller.HasFinalizer(vmCopy, sourceFinalizer) {
+		controller.RemoveFinalizer(vmCopy, sourceFinalizer)
+		vmCopy, err = s.controller.Client.VirtualMachine(vmCopy.Namespace).Update(vmCopy)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	vmCopy.Status.SnapshotInProgress = nil
+	err = s.controller.vmStatusUpdater.UpdateStatus(vmCopy)
+	if err != nil {
+		return true, err
+	}
+
+	return true, nil
+}
+
+func (s *vmSnapshotSource) Spec() snapshotv1.SourceSpec {
+	vmCpy := s.vm.DeepCopy()
+	vmCpy.Status = kubevirtv1.VirtualMachineStatus{}
+	return snapshotv1.SourceSpec{
+		VirtualMachine: vmCpy,
+	}
+}
+
+func (s *vmSnapshotSource) Online() (bool, error) {
+	vmRunning, err := checkVMRunning(s.vm)
+	if err != nil {
+		return false, err
+	}
+
+	exists, err := s.controller.checkVMIRunning(s.vm)
+	if err != nil {
+		return false, err
+	}
+
+	return (vmRunning || exists), nil
+}
+
+func (s *vmSnapshotSource) GuestAgent() (bool, error) {
+	condManager := controller.NewVirtualMachineInstanceConditionManager()
+	vmi, exists, err := s.controller.getVMI(s.vm)
+	if err != nil || !exists {
+		return false, err
+	}
+
+	return condManager.HasCondition(vmi, kubevirtv1.VirtualMachineInstanceAgentConnected), nil
+}
+
+func (s *vmSnapshotSource) Frozen() (bool, error) {
+	vmi, exists, err := s.controller.getVMI(s.vm)
+	if err != nil || !exists {
+		return false, err
+	}
+
+	return vmi.Status.FSFreezeStatus == "frozen", nil
```

Comment:
I added a const variable for "frozen", not sure if in the merged PR or the other PR its in api.FSFROZEN